src/viadot/sources/eurostat.py

```diff
@@ -0,0 +1,356 @@
+
+    def __init__(
+        self,
+        *args,
+        dataset_code: str,
+        params: dict = None,
+        columns: list = None,
+        tests: dict = None,
+        **kwargs
+    ):
+        """It is using HTTPS REST request to pull the data.
+
+        No API registration or API key are required. Data will pull based on parameters
+        provided in dynamic part of the url.
+
+        Example of url: https://ec.europa.eu/eurostat/api/dissemination/statistics/1.0/
+            data/TEIBS020/?format=JSON&lang=EN&indic=BS-CSMCI-BAL
+
+        Static part: https://ec.europa.eu/eurostat/api/dissemination/statistics/1.0/data
+        Dynamic part: /TEIBS020/?format=JSON&lang=EN&indic=BS-CSMCI-BAL
+
+        Please note that for one dataset there are usually multiple data regarding
+        different subjects. In order to retrieve data that you are interested in you
+        have to provide parameters with codes into 'params'.
+
+        Args:
+            base_url (str): The base URL used to access the Eurostat API.
+                This parameter specifies the root URL for all requests made to the API.
+                It should not be modified unless the API changes its URL scheme.
+                Defaults to "https://ec.europa.eu/eurostat/api/dissemination/statistics
+                    /1.0/data/"
+        Args:
+            dataset_code (str): The code of Eurostat dataset that we would like
+                to upload.
+            params (Dict[str], optional):
+                A dictionary with optional URL parameters. The key represents the
+                parameter id, while the value is the code for a specific parameter,
+                for example: params = {'unit': 'EUR'} where "unit" is the parameter
+                that you would like to set and "EUR" is the code of the
+                specific parameter. You can add more than one parameter,
+                but only one code per parameter! So you CAN NOT provide list of codes
+                as in example 'params = {'unit': ['EUR', 'USD', 'PLN']}'
+                These parameters are REQUIRED in most cases to pull a specific
+                dataset from the API. Both parameter and code has to be provided
+                as a string! Defaults to None.
+            columns (List[str], optional): list of needed names of columns.
+                Names should be given as str's into the list. Defaults to None.
+            tests:
+                - `column_size`: dict{column: size}
+                - `column_unique_values`: list[columns]
+                - `column_list_to_match`: list[columns]
+                - `dataset_row_count`: dict: {'min': number, 'max', number}
+                - `column_match_regex`: dict: {column: 'regex'}
+                - `column_sum`: dict: {column: {'min': number, 'max': number}}
+        """
+        self.dataset_code = dataset_code
+        self.params = params
+        self.columns = columns
+        self.tests = tests
```

Comment:
Since these things are attributes of an API call, not of the API instance, they shouldn't be attributes, but rather inputs to relevant methods. Eg. you should be able to call `to_df()` twice and get two different datasets, just by specifying different params. The way this is done here, you need to create a separate class instance for each call.

This would also drastically simplify the code. But it's a bunch of work so probably should just go into the tracker for now.

Reply:
I can not do so, as for now to_df method is inherited form the base Source class and it only takes 2 parameters self adn if_empty. If I will try to add more there will be error